File: Web/Metadata.js
/**
 * Defines the HTML page metadata for search engines.
 */
class SearchEngineMetadata {
    constructor(siteName, locale, title, titlePrefix, titleSuffix, description, image) {
        this.defaults = {
            title: null,
            titlePrefix: null,
            titleSuffix: null,
            robots: SearchEngineMetadata.robots.follow,
            type: SearchEngineMetadata.type.website,
            modifiedTime: null,
            section: null,
            description: null,
            image: null
        };

        this.siteName = siteName;
        this.locale = locale;
        this.titlePrefix = titlePrefix;
        this.titleSuffix = titleSuffix;
        this.description = description;
        this.Image = image;
    }

    static type = {
        article: 'article',
        profile: 'profile',
        website: 'website'
    }

    static robots = {
        follow: 'index, follow',
        nofollow: 'noindex, nofollow'
    }

    /**
     * Default values when creating or resetting metadata.
     */
    static defaults = {

    }

    /**
     * Reset the metadata to default values.
     */
    reset() {
        Object.assign(this, SearchEngineMetadata.defaults);
    }

    /**
     * Creates, updates or removes metadata nodes.
     */
    apply() {
        document.title = [this.titlePrefix, this.title, this.titleSuffix].filterEmpty().join(' ').trim();
        let route = Router.current.route;
        let image = (this.image != null)
            ? Routes.image.uri.relative({ id: this.image, size: 'original' })
            : '/images/logo.png';

        let metadata = [
            { name: 'og:title',         attribute: 'property',  value: document.title },
            { name: 'robots',           attribute: 'name',      value: this.robots },
            { name: 'description',      attribute: 'name',      value: this.description },
            { name: 'og:description',   attribute: 'property',  value: this.description },
            { name: 'og:type',          attribute: 'property',  value: this.type },
            { name: 'og:section',       attribute: 'property',  value: this.section },
            { name: 'og:modified_time', attribute: 'property',  value: this.modifiedTime },
            { name: 'og:image',         attribute: 'property',  value: image },
            { name: 'og:url',           attribute: 'property',  value: (route ? route.uri.canonical(Router.current.params) : null) }
        ];

        document.head.applyMetadata(metadata);
    }
}

File: Web/Metadata.cs
namespace WebCommons.Web
{
    /// <summary>
    /// Defines the HTML page metadata for search engines.
    /// </summary>
    public class SearchEngineMetadata
    {
        public string Title { get; set; } = null;
        public string TitlePrefix { get; set; } = null;
        public string TitleSuffix { get; set; } = null;
        public string FullTitle { get {
                return (this.TitlePrefix + " " + this.Title + " " + this.TitleSuffix).Trim();
            } }

        public string Section { get; set; } = null;
        public string Description { get; set; } = null;
        public string Image { get; set; } = null;
        public string SiteName { get; set; } = null;
        public string Domain { get; set; } = null;
        public string Url { get; set; } = null;
        public string Locale { get; set; } = null;
        public DateTime? ModifiedTime { get; set; } = null;
        public RobotIndexing Robots { get; set; } = RobotIndexing.Follow;
        public enum RobotIndexing { Follow, NoFollow }
        private static readonly Dictionary<RobotIndexing, string> RobotsMap = new Dictionary<RobotIndexing, string>() {
            { RobotIndexing.Follow, "index, follow" },
            { RobotIndexing.NoFollow, "noindex, nofollow" } };
        public PageType Type { get; set; } = PageType.Website;
        public enum PageType { Article, Profile, Website, Other }
        public string Html
        {
            get
            {
                string html = "<title>" + this.FullTitle + "</title>";
                html += this.createMetadataElement("og:title", "property", this.FullTitle);
                html += this.createMetadataElement("description", "name", this.Description);
                html += this.createMetadataElement("og:description", "property", this.Description);
                html += this.createMetadataElement("og:type", "property", (this.Type == PageType.Other) ? null : Enum.GetName(typeof(PageType), this.Type).ToLower());
                html += this.createMetadataElement("og:section", "property", this.Section);
                html += this.createMetadataElement("og:site_name", "property", this.SiteName);
                html += this.createMetadataElement("og:locale", "property", this.Locale);
                html += this.createMetadataElement("og:modified_time", "property", (this.ModifiedTime == null) ? null : this.ModifiedTime.Value.ToString("yyyy-MM-dd"));
                html += this.createMetadataElement("og:image", "property", this.Domain + this.Image);
                html += this.createMetadataElement("robots", "name", RobotsMap[this.Robots]);
                html += this.createMetadataElement("url", "name", this.Domain + this.Url);
                return html;
            }
        }

        private string createMetadataElement(string name, string attribute, string value)
        {
            if (string.IsNullOrEmpty(value)) { return ""; }
            return "<meta " + attribute + "=\"" + name + "\" content=\"" + value + "\" />";
        }
    }
}